node/src/components/storage.rs

```diff
@@ -1186,31 +1200,41 @@ impl StorageInner {
                 responder,
             } => {
                 let mut txn = self.env.begin_rw_txn()?;
-                if !txn.put_value(
-                    self.block_header_db,
-                    &block_header.hash(self.verifiable_chunked_hash_activation),
-                    &block_header,
-                    false,
-                )? {
-                    error!(
-                        ?block_header,
-                        "Could not insert block header (maybe already inserted?)",
-                    );
-                    txn.abort();
-                    return Ok(responder.respond(false).ignore());
+                let block_header_hash = block_header.hash(self.verifiable_chunked_hash_activation);
+                match self.put_block_header(&mut txn, *block_header) {
+                    Ok(result) => {
+                        txn.commit()?;
+                        responder.respond(result).ignore()
+                    }
+                    Err(err) => {
+                        error!(?err, ?block_header_hash, "error when storing block header");
+                        txn.abort();
+                        return Err(err);
+                    }
                 }
-
-                {
-                    let mut indices = self.indices.write()?;
-                    insert_to_block_header_indices(
-                        &mut indices,
-                        &block_header,
-                        self.verifiable_chunked_hash_activation,
-                    )?;
+            }
+            StorageRequest::PutHeadersBatch {
+                block_headers,
+                responder,
+            } => {
+                let mut txn = self.env.begin_rw_txn()?;
+                let mut result = false;
+                for block_header in block_headers.into_iter() {
+                    let block_header_hash =
+                        block_header.hash(self.verifiable_chunked_hash_activation);
+                    match self.put_block_header(&mut txn, block_header) {
+                        Ok(single_result) => {
+                            result = result && single_result;
+                        }
+                        Err(err) => {
+                            error!(?err, ?block_header_hash, "error when storing block header");
+                            txn.abort();
```

Comment:
If we abort, we probably should update the block header index and remove the ones that were added during this transaction? (Also in other places where we call `put_block_header`.)

Reply:
Done in [`7a00eda` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/7a00edae7ec6e8fdc9853c7fd8663b94bb1201a3)